### DISCUSSION TITLE
Can MemoryAllocator set max capacity?

### DISCUSSION THREAD

**Discussion by Yohahaha** [facebookincubator/velox#4137] [Q&A] [answered]
For `MallocAllocator`, its constructor has no params for max capacity, it seems every allocate interface only check capacity by reservationCB. How to check capacity when reservationCB is nullptr?
For `MmapAllocator`, its `capacity_` has min capacity limit with 64 * 256. Can we set smaller capacity?

Is any design docs to describe the relations for `MemoryAllocator`, `MemoryPool`, `MemoryUsageTracker` and `MemoryManager`? They all has maxMemory variable or similar.

cc @xiaoxmeng @mbasmanova

**Comment by xiaoxmeng** [comment, marked as answer]
The memory reservation callback will always be set to enforce memory limit by MemoryPool. Unlike MmapAllocator, MallocAllocator doesn't need to check capacity as it delegates the memory allocations to std::malloc. MmapAllocator manages the mmaped memory space by itself so needs the capacity enforcement.

There is no updated design as for now but we will publish one soon.
